src/aui/auibook.cpp

```diff
@@ -3446,6 +3446,127 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
     }
 }
 
+class wxAuiLayoutObject
+{
+public:
+    wxAuiLayoutObject(wxSize &s, unsigned short l, unsigned short dir, unsigned short r, unsigned short pos)
+    {
+        size = s;
+        layer = l;
+        direction = dir;
+        switch (dir)
+        {
+            case wxAUI_DOCK_CENTER: direction = 0; break;
+            case wxAUI_DOCK_LEFT:   direction = 1; break;
+            case wxAUI_DOCK_RIGHT:  direction = 2; break;
+            case wxAUI_DOCK_TOP:    direction = 3; break;
+            case wxAUI_DOCK_BOTTOM: direction = 4; break;
+            default:                direction = 5;
+        }
+        row = r;
+        position = pos;
+    }
+    void MergeLayout(wxAuiLayoutObject *lo2)
+    {
+        if (this == lo2)
+            return;
+
+        bool mergeHorizontal;
+        if (layer != lo2->layer || direction != lo2->direction)
+            mergeHorizontal = lo2->direction < 3;
+        else if (row != lo2->row)
+            mergeHorizontal = true;
+        else
+            mergeHorizontal = lo2->direction >= 3;
+
+        if (mergeHorizontal)
+        {
+            size.x += lo2->size.x;
+            if (lo2->size.y > size.y) size.y = lo2->size.y;
+        }
+        else
+        {
+            if (lo2->size.x > size.x) size.x = lo2->size.x;
+            size.y += lo2->size.y;
+        }
+    }
+
+    wxSize size;
+    unsigned short layer;
+    unsigned char direction;
+    unsigned short row;
+    unsigned short position;
+
+    static int CompareLayoutObject(wxAuiLayoutObject *lo1, wxAuiLayoutObject *lo2)
+    {
+        int res = lo1->layer - lo2->layer;
+        if (res)
+            return res;
+        res = lo1->direction - lo2->direction;
+        if (res)
+            return res;
+        res = lo1->row - lo2->row;
+        if (res)
+            return res;
+        return lo1->position - lo2->position;
+    }
+};
+
+WX_DEFINE_SORTED_ARRAY(wxAuiLayoutObject*, wxAuiLayoutObjectArray);
```

Comment:
Please let's avoid these macros in the new code. Just use `wxVector<>` with binary search for insertion instead.

Reply:
All right. Didn't know that. I guess I should do the same in my projects? But I used `wxVectorSort` instead of binary search insertion as I found [this reference](https://stackoverflow.com/questions/2710221/is-there-a-sorted-vector-class-which-supports-insert-etc/2710332#2710332) online which states that inserting in a vector might be inefficient as all later elements need to be moved, which sounds reasonable to me. I also guess that most of the layouts aren't that complicated and we won't have many vector elements anyways.